src/hotspot/os/linux/os_linux.cpp

```diff
@@ -233,17 +226,59 @@ julong os::Linux::available_memory() {
     }
     if (mem_limit > 0 && mem_usage > 0) {
       avail_mem = mem_limit > mem_usage ? (julong)mem_limit - (julong)mem_usage : 0;
-      log_trace(os)("available container memory: " JULONG_FORMAT, avail_mem);
-      return avail_mem;
     }
   }
+  return avail_mem;
+}
+
+julong os::available_memory() {
+  return Linux::available_memory();
+}
+
+julong os::Linux::available_memory() {
+  julong avail_mem = available_memory_in_container();
+  if (avail_mem != static_cast<julong>(-1L)) {
+    log_trace(os)("available container memory: " JULONG_FORMAT, avail_mem);
+    return avail_mem;
+  }
 
-  sysinfo(&si);
-  avail_mem = (julong)si.freeram * si.mem_unit;
+  FILE *fp = os::fopen("/proc/meminfo", "r");
+  if (fp) {
```

Comment:
Please don't use implicit booleans. `fp != nullptr` should be fine. See: https://github.com/openjdk/jdk/blob/master/doc/hotspot-style.md#miscellaneous